### DISCUSSION TITLE
How to make changes to `_module.args.pkgs` available to hosts?

### DISCUSSION THREAD

**Discussion by z0al** [hercules-ci/flake-parts#237] [Q&A] [answered]
If I make changes to `pkgs` e.g. adding overlay as shown below:

```nix
{
  perSystem = { pkgs, system, ... }: {
    _module.args.pkgs = import nixpkgs { 
		inherit system; 
		overlays = [ myOverlay ]; };
  	};
}
```

These are not picked up automatically by `lib.nixosSystem` or `lib.darwinSystem` and I get errors when trying to use something from the overlay.

The only way I could get it to work is to use `withSystem` and pass `pkgs` in `specialArgs` but it feels rather hacky:

```nix
 withSystem "x86_64-linux" ({ system, pkgs, ... }:
	inputs.lib.nixosSystem {
		inherit system;
		specialArgs = {
			inherit pkgs inputs; # <-- this
		};
        # ... rest of config omitted 
	}
```

Is that the correct way to do things?

**Comment by roberth** [comment, marked as answer]
IIRC that allows `nixpkgs.*` to be set by modules, which means that some definitions that NixOS modules want to set will be ignored silently.
Ideally NixOS would detect `specialArgs.pkgs` and do this for you, but I'd recommend to import `nixpkgs.nixosModules.readOnlyPkgs` when you do that.

If you do want NixOS modules to further change your Nixpkgs package set, and specifically `config`, you could pass the overlay into `nixpkgs.overlays`. If you only need NixOS modules to set `nixpkgs.overlays` and not `nixpkgs.config`, you could set `nixpkgs.pkgs` + `nixpkgs.overlays`, but that's no more efficient than just passing the overlays without `nixpkgs.pkgs`.

This is way too confusing :(

Things to be improved
- NixOS could detect `specialArgs.pkgs`. Maybe not fail like `readOnlyPkgs`, but warn instead?
- Here, we should have a #74

**Reply by z0al** [reply]
Thanks for the detailed answer.

I imported `nixpkgs.nixosModules.readOnlyPkgs` as you recommended. But yeah, this is indeed super confusing.

> Here, we should have a https://github.com/hercules-ci/flake-parts/issues/74

Sounds exciting.
